### ISSUE TITLE
Enhancement: no-floating-promises shouldn't allow returning a variable containing an unawaited promise. Returning an unawaited promise directly is fine.

### ISSUE THREAD

**Issue by BrannJoly** [typescript-eslint/typescript-eslint#8068]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-floating-promises/

### Description

**no-floating-promise** considers returning an unawaited promise is always ok. 
While this is fine when you return the promise directly (ie `return f()`), this can be an issue if there are some intermediary steps before returning. I just spent a few hours trying to fix such a bug !

For example consider the following pseudocode :

```TypeScript
async createObject(id:string){ ..... }

async x(id:string){
    const a = createObject(id); // note the missing await
    await doSomethingWithObject(id); // will fail because the object doesn't exist yet
    return a; // return the unawaited Promise, eslint doesn't complain
}
```

if doSomethingWithObject logically depends on the promise being on the previous line being resolved, this is a bug and should be flagged as such by eslint!

if doSomethingWithObject doesn't depend on the previous line being resolved, then we can rearrange the code: 
```TypeScript

  async x(){
     await doSomethingWithObject(); //doesn't depend on anything, so can be moved to the first line
     return createObject(); // here, the missing await is ok as we are returning immediately
     // the caller will be responsible for handling the promise properly
  }

```

### Fail

```TypeScript

async f(){ ..... }

async x(){
    const a = f();
    g();
    return a; // mishandled promise. Cannot return a variable containing an unawaited promise. 
    // await it, or return it directly ( ie. return f() ) 
}
```

### Pass

```TypeScript
async f(){ ..... }

async x(){
    g();
    return f(); // g doesn't need to be executed after f, move g() before f() and return f() directly

or 

async f(){ ..... }

async x(){
    const a = await f(); // g is supposed to be executed after f, add the missing await
    g(); 
    return a; 
}
```

**Comment by bradzacher** [comment]
You probably want this rule
https://typescript-eslint.io/rules/return-await/

**Comment by BrannJoly** [comment]
Thank you for the suggestion. Actually, return-await doesn't cover my use-case (I still want to be able to return an unawaited promise at the end of a function)
I would have thought my proposed change would have caught quite a few nasty bugs with no real drawbacks!

Anyway, thank you for the quick turnaround !
